Upgrade Px-Backup utility function

mkoppal-px opened the pull request:
**What this PR does / why we need it**:
To create helper function for upgrade of Px-Backup from the current version to the desired version

- [x] Created helper function `UpgradePxBackup`
- [x] Created helper function `GetPxBackupVersion`
- [x] Created helper function `GetPxBackupVersionSemVer`
- [x] Created helper function `GetPxBackupBuildDate`
- [x] Created helper function `deleteJobAndWait`
- [x] Added a placeholder test `UpgradePxBackup`. This test case is just to verify the upgrade functionality and not to be run in the pipeline

**Which issue(s) this PR fixes** (optional)
Closes #PA-570

**Special notes for your reviewer**:
[Jenkins run - upgrade from 2.3.3 to 2.4.0](https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/594/console)

kphalgun-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+		return err
+	}
+
+	if currentBackupVersion.GreaterThanOrEqual(versionToUpgradeSemVer) {
```
To bypass this check, how about we pass a bool argument like force?

mkoppal-px replied:
We should never bypass this. Upgrade is alway from (n) to (n +x) version. If you want to try it out for dev purposes then just comment out the line.

kphalgun-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	if !strings.EqualFold(postUpgradeVersion, versionToUpgrade) {
+		return fmt.Errorf("expected version after upgrade was %s but got %s", versionToUpgrade, postUpgradeVersion)
+	}
+	log.InfoD("Upgrade from %s to %s is complete", currentBackupVersionString, postUpgradeVersion)
```
How about log.InfoD("Px-Backup upgrade from %s to %s is complete", currentBackupVersionString, postUpgradeVersion)?

mkoppal-px replied:
Done

kphalgun-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+			}
+			return "", false, err
+		}
+		return "", true, fmt.Errorf("job %s not deleted", job.Name)
```
How about fmt.Errorf("job %s not deleted. Err: %v", job.Name, err.Error())?

mkoppal-px replied:
There is no error to return here

kphalgun-px replied:
You are right. My bad.

kphalgun-px commented on tests/backup/backup_upgrade_test.go:
```diff
@@ -0,0 +1,29 @@
+	JustBeforeEach(func() {
+		StartTorpedoTest("UpgradePxBackup", "Upgrading backup", nil, 0)
+	})
+	It("Basic Backup Creation", func() {
```
It("Upgrade Px Backup", func() { ?

mkoppal-px replied:
Fixed

kphalgun-px commented on tests/backup/backup_upgrade_test.go:
```diff
@@ -0,0 +1,29 @@
+		Step("Upgrade Px Backup", func() {
+			log.InfoD("Upgrade Px Backup to version %s", latestPxBackupVersion)
+			err := UpgradePxBackup(latestPxBackupVersion)
+			log.FailOnError(err, "Failed during upgrade")
+			log.InfoD("Upgrade process complete")
```
How about dash.VerifyFatal(err, nil, "Upgrading Px-Backup")?

mkoppal-px replied:
We are using `log.FailOnError` for validating an error. That is a general rule that we need to follow instead of using `dash.VerifyFatal`

mkoppal-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	return nil
+}
+
+// deleteJobAndWait waiting for resources to be deleted
```
Fixed

mkoppal-px commented on tests/backup/backup_upgrade_test.go:
```diff
@@ -0,0 +1,29 @@
+
+// This testcase verifies Px Backup upgrade
+var _ = Describe("{UpgradePxBackup}", func() {
+	var ()
```
Fixed

mkoppal-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	if err != nil {
+		return err
+	}
+	if len(allJobs.Items) > 0 {
```
Sometimes the `pxcentral-post-install-hook` job might not be present, that should not throw any error or stop the upgrade process.

kphalgun-px replied:
How about logging that the job is not present in that case?

mkoppal-px replied:
Fixed

kshithijiyer-px commented on tests/backup/backup_helper.go:
```diff
@@ -58,6 +62,11 @@ const (
 	mongodbPodStatusRetryTime                 = 30 * time.Second
 	backupLocationDeleteTimeout               = 30 * time.Minute
 	backupLocationDeleteRetryTime             = 30 * time.Second
+	latestPxBackupVersion                     = "2.4.0"
```
Would it be a better approach if we could fetch this rom env?

tthurlapati-px replied:
My opinion: if we want a particular commit to always upgrade only to a "specific latest version", then it makes sense to hardcode it (like a particular test being designed for particular version, I guess?). 
But if we want _any_ commit to always upgrade to the latest (or version to upgrade to) as defined by the user (env var), it yeah, env var is better.

mkoppal-px replied:
This is the just an indication of what is the latest version of Px Backup currently going on. User can pass any string for `versionToUpgrade`

kshithijiyer-px replied:
Cool

kshithijiyer-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	storageClassName := pvcs.Items[0].Spec.StorageClassName
+
+	// Get the tarball required for helm upgrade
+
```
Can we remove the blank line here?

mkoppal-px replied:
Done

kshithijiyer-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	log.Infof("Terminal output: %s", output)
+
+	// Execute helm upgrade using cmd
+
```
Can we remove the blank line here?

mkoppal-px replied:
Done

kshithijiyer-px reviewed (commented):
How does this PR handle hop cases?

tthurlapati-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+		return err
+	}
+
+	// Checking if all pods are running
+	allPods, err := core.Instance().GetPods(pxBackupNamespace, nil)
+	log.Infof("Listing all the pods in Px Backup Namespace [%s] - ", pxBackupNamespace)
+	for _, pod := range allPods.Items {
+		log.Infof(pod.GetName())
+	}
+	if err != nil {
+		return err
+	}
+	for _, pod := range allPods.Items {
+		log.Infof("Checking status for pod - %s", pod.GetName())
+		core.Instance().ValidatePod(&pod, 5*time.Minute, 30*time.Second)
+	}
```
Did you check for error later because you're sure that _even if there is an error_, there will _always_ be a list of all pods to go through and that list is useful even if there's an error?
Shouldn't error check be done first?

tthurlapati-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+func GetPxBackupVersion() (string, error) {
+	ctx, err := backup.GetAdminCtxFromSecret()
+	if err != nil {
+		return "", err
+	}
+	versionResponse, err := Inst().Backup.GetPxBackupVersion(ctx, &api.VersionGetRequest{})
+	if err != nil {
+		return "", err
+	}
+	version := versionResponse.GetVersion()
+	return fmt.Sprintf("%s.%s.%s-%s", version.GetMajor(), version.GetMinor(), version.GetPatch(), version.GetGitCommit()), nil
+}
+
+// GetPxBackupVersionSemVer returns the version of Px Backup in semver format like 2.4.0
+func GetPxBackupVersionSemVer() (string, error) {
+	ctx, err := backup.GetAdminCtxFromSecret()
+	if err != nil {
+		return "", err
+	}
+	versionResponse, err := Inst().Backup.GetPxBackupVersion(ctx, &api.VersionGetRequest{})
+	if err != nil {
+		return "", err
+	}
+	version := versionResponse.GetVersion()
+	return fmt.Sprintf("%s.%s.%s", version.GetMajor(), version.GetMinor(), version.GetPatch()), nil
+}
+
+// GetPxBackupBuildDate returns the Px Backup build date
+func GetPxBackupBuildDate() (string, error) {
+	ctx, err := backup.GetAdminCtxFromSecret()
+	if err != nil {
+		return "", err
+	}
+	versionResponse, err := Inst().Backup.GetPxBackupVersion(ctx, &api.VersionGetRequest{})
+	if err != nil {
+		return "", err
+	}
+	version := versionResponse.GetVersion()
+	return version.GetBuildDate(), nil
+}
```
Wouldn't it be cleaner if move getting version into it's own function:
```go
func GetPxBackupVersion() (*api.Version......, error) {
	ctx, err := backup.GetAdminCtxFromSecret()
	if err != nil {
		return "", err
	}
	versionResponse, err := Inst().Backup.GetPxBackupVersion(ctx, &api.VersionGetRequest{})
	if err != nil {
		return "", err
	}
	version := versionResponse.GetVersion()
	return version
}
```
and then have two functions:
```go
func GetPxBackupVersionString() (string, error) {
	version, err := GetPxBackupVersion()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s.%s.%s-%s", version.GetMajor(), version.GetMinor(), version.GetPatch(), version.GetGitCommit()), nil
}
```
and
```go
func GetPxBackupVersionSemVer() (string, error) {
	version, err := GetPxBackupVersion()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s.%s.%s", version.GetMajor(), version.GetMinor(), version.GetPatch()), nil
}
```
and maybe not have `GetPxBackupBuildDate` as it's simply gonna be doing `version.GetBuildDate()` for anyone using it.

mkoppal-px replied:
Fixed based on the first part of your comment but for the second part

`version.GetBuildDate()` is not that intuitive when compared to having a function with name `GetPxBackupBuildDate`
Which is why I thought of going this way.

ak-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+}
+
+// GetPxBackupVersionSemVer returns the version of Px Backup in semver format like 2.4.0
+func GetPxBackupVersionSemVer() (string, error) {
```
can we internally call GetPxBackupVersion() here.

ak-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+
+// GetPxBackupBuildDate returns the Px Backup build date
+func GetPxBackupBuildDate() (string, error) {
+	ctx, err := backup.GetAdminCtxFromSecret()
```
Same as above, code can be replaced with GetPxBackupVersion()

ak-px commented on tests/backup/backup_helper.go:
```diff
@@ -1426,3 +1435,200 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	if err != nil {
+		return err
+	}
+	currentBackupVersion, err := version.NewSemver(currentBackupVersionString)
```
I think we can directly return Semver with GetPxBackupVersionSemVer()

mkoppal-px replied:
But in `GetPxBackupVersionSemVer()` the return type is string, to do `GreaterThanOrEqual` the expected type is `Version`

mkoppal-px commented:
> How does this PR handle hop cases?

You have this run this function twice like say your current version is `2.3.0`
```
UpgradePxBackup("2.3.3")
UpgradePxBackup("2.4.0")
UpgradePxBackup("2.5.0")
```

snigdha-px commented on tests/backup/backup_helper.go:
```diff
@@ -58,6 +62,11 @@ const (
 	backupLocationDeleteRetryTime             = 30 * time.Second
+	latestPxBackupVersion                     = "2.4.0"
+	latestPxBackupHelmBranch                  = "master"
+	pxCentralPostInstallHookJobName           = "pxcentral-post-install-hook"
```
post_install_hook_pod is already declared in portworx.go  as post_install_hook_pod ,although it does not follow camel convention.
Either correct that to camel convention and reuse here or vice-versa

snigdha-px commented on tests/backup/backup_helper.go:
```diff
@@ -1445,3 +1454,205 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	_, err = task.DoRetryWithTimeout(postInstallHookJobCompletedCheck, 10*time.Minute, 30*time.Second)
+	if err != nil {
+		return err
+	}
```
Please check the function ValidateBackupCluster in portworx.go
It does the same verification 
The function ValidateBackupCluster can be modified to add verification like "DoRetryWithTimeout".

lsrinivas-pure reviewed (approved):
lgtm

kphalgun-px reviewed (approved):
LGTM

kshithijiyer-px reviewed (approved):
LGTM

kshithijiyer-px commented:
> > How does this PR handle hop cases?
> 
> You have this run this function twice like say your current version is `2.3.0`
> 
> ```
> UpgradePxBackup("2.3.3")
> UpgradePxBackup("2.4.0")
> UpgradePxBackup("2.5.0")
> ```

Okay looks good. One suggestion I would have if if we can fetch it from some env and run the upgrade multiple times rest looks good to me.

snigdha-px reviewed (approved):
LGTM

tthurlapati-px commented on tests/backup/backup_helper.go:
```diff
@@ -1443,3 +1452,207 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	return fmt.Sprintf("%s.%s.%s", backupVersion.GetMajor(), backupVersion.GetMinor(), backupVersion.GetPatch()), nil
+}
+
+// GetPxBackupBuildDate returns the Px Backup build date
+func GetPxBackupBuildDate() (string, error) {
+	ctx, err := backup.GetAdminCtxFromSecret()
+	if err != nil {
+		return "", err
+	}
+	versionResponse, err := Inst().Backup.GetPxBackupVersion(ctx, &api.VersionGetRequest{})
+	if err != nil {
+		return "", err
+	}
+	backupVersion := versionResponse.GetVersion()
+	return backupVersion.GetBuildDate(), nil
+}
```
You can use `GetPxBackupVersion` directly and then do `backupVersion.GetBuildDate`, like in other functions `GetPxBackupVersion***`.

tthurlapati-px commented on tests/backup/backup_helper.go:
```diff
@@ -1443,3 +1452,207 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+		return err
+	}
+
+	// Checking if all pods are running
+	allPods, err := core.Instance().GetPods(pxBackupNamespace, nil)
+	log.Infof("Listing all the pods in Px Backup Namespace [%s] - ", pxBackupNamespace)
+	for _, pod := range allPods.Items {
+		log.Infof(pod.GetName())
+	}
+	if err != nil {
+		return err
+	}
+	for _, pod := range allPods.Items {
+		log.Infof("Checking status for pod - %s", pod.GetName())
+		err = core.Instance().ValidatePod(&pod, 5*time.Minute, 30*time.Second)
+		if err != nil {
+			return err
+		}
+	}
```
Shouldn't error check be done first?
And maybe combine the two loops? I'm not sure why they are two separate loops.
I mentioned it in https://github.com/portworx/torpedo/pull/1237#discussion_r1146042879, and you said you "fixed it". does that mean this is how it _should_ be?

mkoppal-px replied:
Fixed it. We did not need the first loop since we are printing the name of the pods in the second loop anyway.

vpinisetti-px commented on tests/backup/backup_helper.go:
```diff
@@ -1443,3 +1452,207 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+
+	// Execute helm upgrade using cmd
+	log.Infof("Upgrading Px-Backup version from %s to %s", currentBackupVersionString, versionToUpgrade)
+	cmd = fmt.Sprintf("helm upgrade px-central px-central-%s.tgz --namespace %s --create-namespace --version %s --set persistentStorage.enabled=true,persistentStorage.storageClassName=\"%s\",pxbackup.enabled=true",
```
@mkoppal-px --create-namespace  should not be passed with upgrade command. --create-namespace will be used to create a namespace for fresh installation case.

mkoppal-px commented on tests/backup/backup_helper.go:
```diff
@@ -1443,3 +1452,207 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	log.Infof("Terminal output: %s", output)
+
+	// Execute helm upgrade using cmd
+	log.Infof("Upgrading Px-Backup version from %s to %s", currentBackupVersionString, versionToUpgrade)
```
We can consider this once we have NFS GA

vpinisetti-px replied:
Okay..

mkoppal-px commented on tests/backup/backup_helper.go:
```diff
@@ -1443,3 +1452,207 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+
+	// Execute helm upgrade using cmd
+	log.Infof("Upgrading Px-Backup version from %s to %s", currentBackupVersionString, versionToUpgrade)
+	cmd = fmt.Sprintf("helm upgrade px-central px-central-%s.tgz --namespace %s --create-namespace --version %s --set persistentStorage.enabled=true,persistentStorage.storageClassName=\"%s\",pxbackup.enabled=true",
```
Good catch

mkoppal-px commented on tests/backup/backup_helper.go:
```diff
@@ -1443,3 +1452,207 @@ func DeleteBackupAndWait(backupName string, ctx context.Context) error {
+	// Execute helm upgrade using cmd
+	log.Infof("Upgrading Px-Backup version from %s to %s", currentBackupVersionString, versionToUpgrade)
+	cmd = fmt.Sprintf("helm upgrade px-central px-central-%s.tgz --namespace %s --create-namespace --version %s --set persistentStorage.enabled=true,persistentStorage.storageClassName=\"%s\",pxbackup.enabled=true",
+		versionToUpgrade, pxBackupNamespace, versionToUpgrade, *storageClassName)
```
Makes sense. Created a function `ValidateAllPodsInPxBackupNamespace` and performing this check just before running helm update and also after the post hook install job is completed.

mkoppal-px commented:
Latest run with all the addressed comments - https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/677/console
